src/plugins/discover/server/locator/searchsource_from_locator.ts

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { SearchSource, TimeRange } from '@kbn/data-plugin/common';
+import { DataView } from '@kbn/data-views-plugin/common';
+import { AggregateQuery, Filter, Query } from '@kbn/es-query';
+import { LocatorServicesDeps, SavedSearchObjectType } from '.';
+import { DiscoverAppLocatorParams } from '../../common';
+import { getSortForSearchSource } from '../../common/utils/sorting';
+import { getColumns } from './columns_from_locator';
+import { getSearchObjects } from './lib/get_search_objects';
+
+// Shortcut for return type of searchSource.getField('filter');
+type FilterResponse = undefined | Filter | Filter[] | (() => Filter | Filter[] | undefined);
+
+// flattens filter objects coming from different sources
+function normalizeFilter(savedSearchFilterTmp?: FilterResponse) {
+  let savedSearchFilter: Filter[] | undefined;
+  if (savedSearchFilterTmp && Array.isArray(savedSearchFilterTmp)) {
+    // can not include functions: could be recursive
+    savedSearchFilter = [...savedSearchFilterTmp.filter((f) => typeof f !== 'function')];
+  } else if (savedSearchFilterTmp && typeof savedSearchFilterTmp !== 'function') {
+    savedSearchFilter = [savedSearchFilterTmp];
+  }
+  return savedSearchFilter;
+}
+
+/*
+ * Combine the time range filter from the job request body with any filters that have been saved into the saved search object
+ * NOTE: if the filters that were saved into the search are NOT an array, it may be a function, and can not be supported.
+ */
+const getFilters = (
+  timeFieldName: string | undefined,
+  index: DataView,
+  savedSearch: SavedSearchObjectType,
+  searchSource: SearchSource,
+  params: DiscoverAppLocatorParams
+) => {
+  const filters: Filter[] = [];
+
+  // Set a time range filter from (1) DiscoverAppLocatorParams or (2) SavedSearchObjectType
+  if (timeFieldName) {
+    const timeRange = params.timeRange
+      ? params.timeRange
+      : savedSearch.attributes.timeRange
+      ? (savedSearch.attributes.timeRange as TimeRange)
+      : null;
+
+    if (timeRange) {
+      filters.push({
```

Comment:
I didn't use that here, because 
1. The `TimefilterSetup` interface depends on browser-side types.
1. We would have to know in advance whether the time range is absolute or relative. However, the DiscoverAppLocatorParams do not have that information.
 

I'm not sure what extra safety we would get if we were able to use `data.query.timefilter.timefilter` utilities here.